### PR TITLE
feat(dart/catalyst_cardano_serialization)!: Implement tiered fee calculation for reference scripts 

### LINKED ISSUE #870: [Feature]: Implement tiered fee calculation for the Conway era
### Feature Title

Tiered fee calculation for reference scripts

### Feature Description

The Cardano blockchain recently introduced a tiered fee calculation for transactions involving reference scripts with the Conway era (Chang Hardfork). 

This task is to implement the same tiered fee calculation mechanism in the `catalyst_cardano_serialization` package

### Use Case

This modification is necessary to align with the changes in the Cardano Conway era, ensuring that reference script sizes are properly accounted for in transaction fees. 

It improves the security and performance of the blockchain by preventing potential abuse of large reference scripts while maintaining reasonable fees for most DApp transactions.

### Is this a modification to an existing feature?

Yes

### PR DESCRIPTION
**Pull request by ilap**
# Description

This pull request implements the tiered fee calculation model introduced in the Cardano Conway era (Chang Hardfork).

## Related Issue(s)

Closes #870

## Description of Changes

- Replaced the existing LinearFee model with a TieredFee model that applies progressively higher fees for reference script sizes beyond the initial 25KiB.
- Renamed `minNoScriptFee()` to `minFee()` for better clarity.
- Added a method to calculate the length of reference scripts, contributing to the overall fee calculation.
- Introduced the `ReferenceScriptSizeLimitExceededException` class to handle cases where the reference script size exceeds defined limits.
- Added tests for the TieredFee class to ensure that it correctly calculates fees for transactions with reference scripts.
- Added unit tests for `length` calculations and fee calculations involving reference scripts.

## Breaking Changes

- The renaming of `minNoScriptFee()` to `minFee()` may require adjustments in dependent modules that reference this method. 
- The `inputs` field has been updated from `List<TransactionUnspentOutput>` to `Set<TransactionUnspentOutput>`. This may require adjustments in dependent code that expects a List instead of a Set.

Other than that, there are no significant breaking changes.

## Please confirm the following checks

* [x] My code follows the style guidelines of this project
* [x] I have performed a self-review of my code
* [x] I have commented my code, particularly in hard-to-understand areas
* [x] I have made corresponding changes to the documentation
* [x] My changes generate no new warnings
* [x] I have added tests that prove my fix is effective or that my feature works
* [x] New and existing unit tests pass locally with my changes
* [x] Any dependent changes have been merged and published in downstream module

### CODE REVIEW

**Comment by dtscalac** [comment]
@ilap Thanks for the PR. Make sure to denote in the PR's title that it contains breaking changes. I updated this PR title.

File: catalyst_voices_packages/catalyst_cardano_serialization/lib/src/builders/transaction_builder.dart [hunk 1]
```diff
@@ -19,7 +19,7 @@ final class TransactionBuilder extends Equatable {
   ///
   /// Enough [inputs] must be provided to be greater or equal
   /// the amount of [outputs] + [fee].
-  final List<TransactionUnspentOutput> inputs;
+  final Set<TransactionUnspentOutput> inputs;
```
Comment (dtscalac) [inline catalyst_voices_packages/catalyst_cardano_serialization/lib/src/builders/transaction_builder.dart:22]: Please add this to breaking changes. Yes, it's very minor but we don't graduate breaking changes, if a change is breaking then it should be listed.

File: catalyst_voices_packages/catalyst_cardano_serialization/lib/src/fees.dart [hunk 2]
```diff
@@ -1,35 +1,165 @@
...
+            _ => 0,
+          },
+    );
+
+    return totalSize;
   }
 
   @override
-  List<Object?> get props => [constant, coefficient];
+  List<Object?> get props => [constant, coefficient, refScriptByteCost];
```
Comment (dtscalac) [inline catalyst_voices_packages/catalyst_cardano_serialization/lib/src/fees.dart:164]: Please add all missing fields: multiplier, sizeIncrement, maxRefScriptSize.

The `props` are used for hashing/data equality.

**Comment by dtscalac** [comment]
@ilap The changes in `catalyst_cardano_serialization` have caused compile errors in other dependent packages, i.e. `catalyst_cardano`. These packages provide different functionalities, they build on top of catalyst_cardano_serialization but in short, we use them as a test for backward compatibility when changing the serialization package.

Would it be possible to adjust these compilation issues as well?

File: catalyst_voices_packages/catalyst_cardano_serialization/lib/src/fees.dart (hunk 2 above)
Comment (ilap) [reply catalyst_voices_packages/catalyst_cardano_serialization/lib/src/fees.dart:164]: Thank you for pointing that out. 
Initially, I added those fields as static constants, but later decided to include them as constructor parameters, but I missed updating the props list.

**Review by dtscalac — APPROVED** [review]
LGTM.

@stevenj Please support on merging the PR, I validated it successfully.

**Comment by stevenj** [comment]
@ilap Thank you again for the great contribution.

**Comment by ilap** [comment]
Hi @stevenj and @dtscalac,

Could you please update the version of the catalyst_cardano_serialization package and any related dependencies, and publish them on pub.dev?

Thanks in advance!

Cheers,
Pal

**Comment by dtscalac** [comment]
> Hi @stevenj and @dtscalac,
> 
> Could you please update the version of the catalyst_cardano_serialization package and any related dependencies, and publish them on pub.dev?
> 
> Thanks in advance!
> 
> Cheers, Pal

Hi, thanks for reaching out. You can find  the newly updates packages here: [pub.dev](https://pub.dev/publishers/projectcatalyst.io/packages.)

**Comment by dtscalac** [comment]
@ilap can you share your sample? Was it a web project or android/ios?

**Comment by ilap** [comment]
> @ilap can you share your sample? Was it a web project or android/ios?

Hey @dtscalac,

We should continue this discussion in [the issue](#1471)

**Pull request merged**